libraries/base/Java/Primitive.hs

```diff
@@ -57,6 +62,22 @@ instance Data Short where
 
 deriving instance Typeable JChar
 
+charToJChar :: Char -> Maybe JChar
+charToJChar chr
+    | minJChar <= chrVal && chrVal <= maxJChar = Just $ fromIntegral chrVal
+    | otherwise = Nothing
+    where minJChar = fromIntegral (minBound :: JChar)
+          maxJChar = fromIntegral (maxBound :: JChar)
+          chrVal = ord chr
+
+jcharToChar :: JChar -> Maybe Char
+jcharToChar jchr
+    | minChar <= jchrVal && jchrVal <= maxChar = Just $ chr jchrVal
+    | otherwise = Nothing
+    where minChar = ord (minBound :: Char)
+          maxChar = ord (maxBound :: Char)
+          jchrVal = fromIntegral jchr
+
```

Comment:
`JChar` is a 16-bit unsigned number and `Char` is 32-bit. So the condition that it falls within the founds of `Char` will trivially hold - no need to check for bounds. What you do need to check for is whether the `JChar` is a surrogate - if it's a surrogate, then return `Nothing`.

For more details on surrogates, you can take a look [here](https://en.wikipedia.org/wiki/UTF-16).